Add CI Support

ozcodes opened the pull request:
Building using CI for linux x86 and macOS arm uploading the build to github actions and to the release page.
The binaries are signed for macOS except for `rustc` binary which currently cause a failure with cargo build.

Main changes:
- Created a CI for building on linux and macos.
- Created a new bash script for signing macos binaries.
- Changed `package.sh` script to include signing the binaries just before creating the compressed file.
- Changed `justfile` to include update cargo in the build command.

ozcodes committed:
first build test

ozcodes committed:
install just

ozcodes committed:
try using brew

ozcodes committed:
use just all

ozcodes committed:
install ninja

ozcodes committed:
add cargo update

ozcodes committed:
print dir and upload artifact

ozcodes committed:
use relative path

ozcodes committed:
separated actions

ozcodes committed:
code sign llvm-objcopy

ozcodes committed:
sign multiple files

ozcodes committed:
fix secrets

ozcodes committed:
add linux support

ozcodes committed:
install just

ozcodes committed:
run linux only for testing

ozcodes committed:
try with sudo

ozcodes committed:
apt update

ozcodes committed:
install just with dedicated action

ozcodes committed:
install ninja

ozcodes committed:
fix if

ozcodes committed:
update cargo

ozcodes committed:
fix path

ozcodes committed:
fix loop and sign script path

ozcodes committed:
sign rust lib

ozcodes committed:
temporary remove sign.sh

ozcodes committed:
restore sign.sh

ozcodes committed:
no quotation marks

ozcodes committed:
run both macos and linux

ozcodes committed:
[skip ci] remove comment

caballa reviewed (commented):
We do not want to include any binary or .tar.bz2 file in the repo.
Instead, we put them in [Releases](https://github.com/Certora/certora-solana-platform-tools/releases)

ozcodes commented:
> We do not want to include any binary or .tar.bz2 file in the repo. Instead, we put them in [Releases](https://github.com/Certora/certora-solana-platform-tools/releases)

The ci doesn't upload the the files to the repository, its uploaded to the github actions, you can see here an example:
https://github.com/Certora/certora-solana-platform-tools/actions/runs/11880630885
should it be removed from there ?

From the comment i uderstand that we need to upload the files when we create a new release\tag, i'll add support for it in the ci.

update:
@caballa i didn't notice that i added a .tar.bz2 file by mistake to the repo, i removed it, thanks for noticing.

ozcodes committed:
create a release with artifacts

ozcodes committed:
disable binaries signing for macos temporary

ozcodes committed:
restore signing and remove zip file from root dir

ozcodes committed:
disable signing temporary

ozcodes committed:
restore signing and add rust-lld

ozcodes committed:
don't sign rustc

ozcodes committed:
sign other rust bins

ozcodes committed:
remove test branch

ozcodes committed:
fix rust-lld filename

ozcodes committed:
test build

ozcodes committed:
remove test branch

1arie1 commented on justfile:
```diff
@@ -30,11 +30,11 @@ build-cargo:
 [linux]
 build-cargo:
-	cd {{ out_dir }}/cargo && env OPENSSL_STATIC=1 OPENSSL_LIB_DIR=/usr/lib/x86_64-linux-gnu OPENSSL_INCLUDE_DIR=/usr/include/openssl cargo build --release
+	cd {{ out_dir }}/cargo && cargo update && env OPENSSL_STATIC=1 OPENSSL_LIB_DIR=/usr/lib/x86_64-linux-gnu OPENSSL_INCLUDE_DIR=/usr/include/openssl cargo build --release
```
why is `cargo update` needed? This will update `Cargo.lock`, leading to inconsistent builds. It is strange if it was required for the CI but not for me locally.

ozcodes replied:
yes, it was required by the ci.
this is the error from the ci build:
```
Compiling time v0.3.29
error[E0282]: type annotations needed for `Box<_>`
  --> /Users/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/time-0.3.29/src/format_description/parse/mod.rs:83:9
   |
83 |     let items = format_items
   |         ^^^^^
...
86 |     Ok(items.into())
   |              ---- type must be known at this point
   |
   = note: this is an inference error on crate `time` caused by an API change in Rust 1.80.0; update `time` to version `>=0.3.35` by calling `cargo update`
``` 

maybe we can replace cargo update with just updating `time` crate version ?

agurfinkel replied:
Right. I remember getting this error. The issue is that we are compiling old `cargo` using new `cargo` and there some library no longer works. The proper fix is to update `time` package to newer version and apply patch to `Cargo.lock`. I have to do it locally, check that it works, and then commit the patch to the repo. 

Current workaround of calling `cargo update` each time is ok for now

1arie1 commented on scripts/package.sh:
```diff
@@ -82,6 +82,26 @@ if [[ "${HOST_TRIPLE}" != "x86_64-pc-windows-msvc" ]] ; then
 fi
 
+# Sign macOS binaries
+if [[ $HOST_TRIPLE == *apple-darwin* ]]; then
```
this has to be outside of `package.sh` and has to be optional. 

In the current state, I will not be able to build this locally since signing will fail.

ozcodes replied:
your right it will failed locally.
i tried to keep it outside `package.sh` at first but eventually added it there because we need to sign the binaries right after we copy all of binaries to a single folder (`deploy` dir) and before we zip them.

how about adding a flag so the signing process will not run by default but only when ci is running ?
and also make the root directory of the binaries as a variable ?
this way we will have control over it in local builds.

agurfinkel replied:
I think you already use some local variables as part of singing process to communicate the key. You can just check whether they are set and not sign if they are not.

1arie1 commented on scripts/package.sh:
```diff
@@ -82,6 +82,26 @@ if [[ "${HOST_TRIPLE}" != "x86_64-pc-windows-msvc" ]] ; then
+        "$RUST_LIB/librustc_driver-b4e91886a4c059a0.dylib" \
+        "$RUST_LIB/libstd-6eff127b55c063c2.dylib" \
+        "$RUST_LIB_BIN/rust-lld"
+       # "$RUST_BIN/rustc" # Not signing 'rustc' duo to failing cargo build
```
perhaps it is safer to not sign anything until we understand what is the cause of the problem. 
Perhaps signing other binaries creates a problem we just don't see yet.

ozcodes replied:
it might be the case, we are signing the gambit binary for some time now without any issues, on the other hand i'm not sure its the right comparison.
we can comment it out for now until we feel more comfortable with it of course.

agurfinkel replied:
I had people use the unsigned binaries from CI and I know that they work well. I am worried that if signed binaries have some problem that only appears under some conditions, it will be hard to debug since the developers do not use the same binaries and thus will not be able to reproduce the problem. 

Since until we are able to sign all binaries, we still will give instructions on how to manually quarantine using xattr, it is safer to not sign.

I will try to setup local signing key and see if I can reproduce the problem using local build and also look if there are instructions on how to sign the rustc.

ozcodes replied:
ok, i will disable signing for now.
can you share more details on the instructions ? do you mean that we are removing the quarantine attribute from the binaries in order to bypass macos gatekeeper ?

agurfinkel replied:
specifically, we ask users to do this
```
sudo xattr -rd com.apple.quarantine $HOME/platform-tools-certora
```

ozcodes committed:
sign only when running ci and disable signing for now

ozcodes committed:
use cargo 1.75

ozcodes committed:
add macos intel build

ozcodes committed:
add mac intel to release

ozcodes committed:
add toolchain for v1.75

ozcodes committed:
cahnge default toolchain

ozcodes committed:
use version 1.75 and fix default toolchain

ozcodes committed:
use macos-13 for intel
